# Requires Promise-like values to be handled appropriately (`no-floating-promises`)

This rule forbids usage of Promise-like values in statements without handling
their errors appropriately. Unhandled promises can cause several issues, such
as improperly sequenced operations, ignored Promise rejections and more. Valid
ways of handling a Promise-valued statement include `await`ing, returning, and
either calling `.then()` with two arguments or `.catch()` with one argument.

## Rule Details

Examples of **incorrect** code for this rule:

```ts
const promise = new Promise((resolve, reject) => resolve('value'));
promise;

async function returnsPromise() {
  return 'value';
}
returnsPromise().then(() => {});

Promise.reject('value').catch();

Promise.reject('value').finally();
```

Examples of **correct** code for this rule:

```ts
const promise = new Promise((resolve, reject) => resolve('value'));
await promise;

async function returnsPromise() {
  return 'value';
}
returnsPromise().then(
  () => {},
  () => {},
);

Promise.reject('value').catch(() => {});

Promise.reject('value').finally(() => {});
```

## Options

The rule accepts an options object with the following properties:

```ts
type Options = {
  // if true, checking void expressions will be skipped
  ignoreVoid?: boolean;
<<<<<<< HEAD
<<<<<<< HEAD
  // if true, checking for async iife will be skipped
  ignoreIIFE?: boolean;
=======
>>>>>>> 782567c486993431d88a7d42ed2c18702ecdfd4f
=======
  // if true, checking for async iife will be skipped
  ignoreIIFE?: boolean;
>>>>>>> b9690f13ae0b02552662ffbd680e2069f9283c9e
};

const defaults = {
  ignoreVoid: false,
};
```

### `ignoreVoid`

<<<<<<< HEAD
<<<<<<< HEAD
This allows you to stop the rule reporting promises consumed with void operator.
This can be a good way to explicitly mark a promise as intentionally not awaited.
=======
This allows to easily suppress false-positives with void operator.
>>>>>>> 782567c486993431d88a7d42ed2c18702ecdfd4f
=======
This allows you to stop the rule reporting promises consumed with void operator.
This can be a good way to explicitly mark a promise as intentionally not awaited.
>>>>>>> b9690f13ae0b02552662ffbd680e2069f9283c9e

Examples of **correct** code for this rule with `{ ignoreVoid: true }`:

```ts
async function returnsPromise() {
  return 'value';
}
void returnsPromise();

void Promise.reject('value');
```

<<<<<<< HEAD
<<<<<<< HEAD
=======
>>>>>>> b9690f13ae0b02552662ffbd680e2069f9283c9e
### `ignoreIIFE`

This allows you to skip checking of async iife

Examples of **correct** code for this rule with `{ ignoreIIFE: true }`:

```ts
await(async function () {
  await res(1);
})();

(async function () {
  await res(1);
})();
```

## When Not To Use It

If you do not use Promise-like values in your codebase, or want to allow them to remain unhandled.
<<<<<<< HEAD
=======
## When Not To Use It

If you do not use Promise-like values in your codebase or want to allow them to
remain unhandled.
>>>>>>> 782567c486993431d88a7d42ed2c18702ecdfd4f
=======
>>>>>>> b9690f13ae0b02552662ffbd680e2069f9283c9e

## Related to

- TSLint: ['no-floating-promises'](https://palantir.github.io/tslint/rules/no-floating-promises/)
