# Prefer usage of `as const` over literal type (`prefer-as-const`)

This rule recommends usage of `const` assertion when type primitive value is equal to type.

## Rule Details

Examples of **incorrect** code for this rule:

```ts
let bar: 2 = 2;
let foo = <'bar'>'bar';
let foo = { bar: 'baz' as 'baz' };
```

Examples of **correct** code for this rule:

```ts
let foo = 'bar';
let foo = 'bar' as const;
let foo: 'bar' = 'bar' as const;
let bar = 'bar' as string;
let foo = <string>'bar';
let foo = { bar: 'baz' };
```

## When Not To Use It

<<<<<<< HEAD
<<<<<<< HEAD
If you are using TypeScript < 3.4
=======
If you are using typescript < 3.4
>>>>>>> 782567c486993431d88a7d42ed2c18702ecdfd4f
=======
If you are using TypeScript < 3.4
>>>>>>> b9690f13ae0b02552662ffbd680e2069f9283c9e
